cmd/bitcask: recovery tool

jsign opened the pull request:
In the #78 direction.
This is still a WIP since I've made multiple changes and would like second opinions.

1- new `recover` command on `cmd/bitcask`
This is an approximation where I'd prefer some early comments to finish it up.

Discussion 1.a: database configuration: _options_.
Now are hardcoded to default values, which is not correct. Ok, so why not calling `Open` that `getConfig`?. Two reasons: first, this command isn't interested in really doing all the work of opening the database; second, even if is done there isn't an API to get the read configuration.

I think an option would be making `getConfig` exported, but move it away from the `bitcask` package to not really expose it to clients; moving it to some `internal` subpackage that can be used within the project. Opinions?

Discussion 1.b: an `output` flag
I didn't create an output flag to choose the path and name of the recovered index file but just chose `recovered_index` as the default one. Opinions? One flag per possible output (in the future two files will be analyzed, datafile and index file).

2- refactored `internal` package
I considered that the `internal` package had many functions sharing the same name scope which felt quite odd. I created two `internal` sub-packages: `data` and `index`, leaving in the base `internal` package the models and utils/version etc. 

I think this makes more sense to not mix unexported things from `codec_index` with other parts of unrelated code in `internal`. We can rollback this, but I took advantage of and make this proposal. Sounds convincing?

I'll keep progress freezed until a round of discussion, to adjust things and close it up later for a final review.

prologic commented:
Some comments on the discussion points...

1a) I'm not sure about the `Open()` logicists and options/configuration. I'm open to suggestions; but exposing this asn an exported API somehow feels wrong to me. What problems are we trying to address here?

1b) Having flags for where to put recovered files seems like we're asking the user to think more? Maybe just default to a `.recovered` suffix? Or even a `.recovered` database directory for the entire recovered database that a user can then inspect with `bitcask` and/or the library?

2) I saw your refactoring. I _think_ its great and aactually adds a lot of quality imrovements. Thank you!

Let's keep going with this :) I thin kwe're on to a good thing here! Data durability and recovery I think are important -- We have tools for backup/import so we're getting close to ensuring some level of data saftey for any user/use-case.

jsign commented:
1a) Exactly, that's why I mentioned moving the `getConfig` function to an [`internal` package](https://docs.google.com/document/d/1e8kOo3r51b2BWtTs_1uADIA5djfXhPT36s6eHVRIvaU/edit). There we can export `getConfig` without exposing it to bitcask clients, but yes to internal project packages. (one of the advantages of a package being under `internal`).
The need for this is having the `maxKeySize` which is required to read the index file and do its corruption validations.

2b) Cool, currently is done this way.

2. Yeah, I think more structure under the `internal` package its better to have particular namespaces for different things. If not, a `read` function in `internal` isn't clear if is in the `datafile`, `index`, `configuration` scope. I prefer moving files in separate PRs since it becomes a pain to rebase or merge, so I'll maybe do the minimum necessary now, and we can keep improving.

I'll do some minor tweaks and notify where there's a canddate version to review to merge.

jsign commented:
I think we have a reasonable version to review.

prologic commented:
Ool I’ll decide soon!
Also agree on 1)
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

prologic commented on cmd/bitcask/recover.go:
```diff
@@ -0,0 +1,76 @@
+func init() {
+	RootCmd.AddCommand(recoveryCmd)
+	recoveryCmd.Flags().BoolP("dry-run", "n", false, "Will only check files health without applying recovery if unhealthy")
+	viper.BindPFlag("dry-run", recoveryCmd.Flags().Lookup("dry-run"))
```
Best to do this in the `PreRun()` hook; I've run into problems doing this anywhere else

jsign replied:
You mean only the `viper.BindPFlag` line? That would make sense to avoid kidnapping `dry-run` name for `viper` if another command also has that flag?

prologic replied:
Yeap I tend to put the viper binding statements in Viper's `PreRun()` hook

jsign replied:
Cool. I already did that exact change in the pushed commit below.

jsign commented:
Well, an extra commit with changes regarding extracting configuration to an internal sub-package:
- Moved configuration to own internal sub-package.
- I noticed that for serialization and deserialization of configuration, anonymous structs that where mirrors of `config` struct where used. I guess always all the configuration should be considered in serialization and deserialization (if not a `'-'` could be use for ignoring). This was for a reason?
- I renamed `MarshalJSON` and `getConfig` now to simply `Encode` and `Decode` respectively. I guess if it is JSON, or other format is quite unnecessary to made explicit to callers.
- Note: now `Config` seem exported but remember is in `internal`, so the biggest scope of visibility is this project, and not clients.
- Made `DefaultSync` default sync option explicit (instead of implicit)

prologic commented:
> I noticed that for serialization and deserialization of configuration, anonymous structs that where mirrors of config struct where used. I guess always all the configuration should be considered in serialization and deserialization (if not a '-' could be use for ignoring). This was for a reason?

Because originally we didn't want to export the `config` struct :)
